NASA_Challenge_[Barisyaz]_[Ingenuity Flight Simulator]

BarisYazici opened the pull request:
This project simulates the Ingenuity Mars Helicopter using Gazebo and ROS 2 in Space-ROS.

closes #67 

Aim of this project is to develop tools to mimic the Martian environment as close as possible. This enables space robots to easily iterate over the development cycles. This is important especially for the aerial vehicles. Because the air is much thinner and the nature events are different. Having this in mind, I develop the Ingenuity Flight Simulator, calculating the lift and drag forces as close to the Martian atmosphere and gravity as possible. This could enable Nasa engineers to develop controllers that can react to the nature events or test flight controllers with sensor noise such as IMU, altimeter or Camera.

https://github.com/user-attachments/assets/f366b00c-be01-40e3-84ab-def0b843cbdf

https://github.com/user-attachments/assets/4fe554d1-a14e-435e-b8b9-331ad4a1ba7b


https://github.com/user-attachments/assets/26c24000-aaf6-4661-b7e6-0a1d4f1c31c8


![martian_helicopter_top](https://github.com/user-attachments/assets/6be2d1e4-de2e-441e-bf17-e7c48d80ae7c)


```mermaid
graph TD
    A[User Input] -->|/alpha_c, /alpha_s| D[Flight Simulation Plugin]
    A -->|/desired_altitude| C[Flight Control Plugin]
    C -->|/angle_of_attack| D
    D -->|LIft/ Drag Forces & Torques| E[Gazebo Physics Engine]
    E -->|Physics Update| F[Ingenuity Model]
    F -->|State Feedback| E
    E -->|Model State| D
    D -->|Telemetry| G[Visualization]
    
    subgraph Gazebo Simulation
    E
    F
    end
    
    subgraph Gazebo Plugins
    C
    D
    end
    
    classDef userInput fill:#f9f,stroke:#333,stroke-width:2px;
    classDef gazebo fill:#cfc,stroke:#333,stroke-width:2px;
    classDef spaceROS fill:#fcf,stroke:#333,stroke-width:2px;
    classDef visualization fill:#ccf,stroke:#333,stroke-width:2px;
    
    class A userInput;
    class E,F gazebo;
    class C,D spaceROS;
    class G visual

BarisYazici commented:
Thank you for your feedback. I will update the PR based on your comments. I have some questions and suggestions:

###  Feature Separation

I have implemented multiple features in this PR:

- Realistic helicopter simulation
- Automatic altitude control with PID for the collective input of simulation
- Realistic Mars environment mesh and rocks
- Mars dust storm simulation
- Sensor support:
  - Altimeter
  - RGB Camera
  - IMU
- Realistic wind support

Would it make sense to separate these into individual issues and reference them in separate commits?

### Build Script and Docker Image

Regarding the build script, there's a potential issue with the current workflow. My script relies   ‘openrobotics/space_robots_demo’  that comes as an artifact from the docker repo under the space-ros project. To address this, I have two questions:

1. Should I base my Dockerfile on Space-ROS's publicly available Docker image?
2. Is there a plan to make the ‘openrobotics/space_robots_demo’ image I'm currently basing on available on Docker Hub?

BarisYazici commented:
Okay I will update it with space-ros docker image. I haven’t pushed anything new yet. Could you also tell me if I should open multiple issues for multiple features or one issue listing multiple features?

gmacario commented:
> Okay I will update it with space-ros docker image. I haven’t pushed anything new yet. Could you also tell me if I should open multiple issues for multiple features or one issue listing multiple features?

According to @mkhansenbot comment in https://github.com/space-ros/demos/pull/39#issuecomment-2334863879

> Since this is an all new demo, I don't think that's necessary. Just document what these capabilities are in your README, etc.

BarisYazici committed:
resolves #67 support ingenuity realistic flight simulation and control

fix: docker container support

BarisYazici commented:
> Thanks for your submission, however this doesn't build in our CI, because you're relying on the 'space_robots' docker image, which you're not building. So there are two options:
> 
> 1. build on the space-ros:latest image only, pull in any assets you need from the space-ros/simulation repo
> 2. Implement a multi-stage build where you first build the `space_robots` image then you build yours on top of that
> 
> I recommend option 1 if possible, as the space_robots demo will be refactored and moved into this repo later, so it's preferable to remove that dependency if you can

Thanks @mkhansenbot for the update. I updated the dockerfile to be based on space-ros public image and the build script. Workflow should hopefully work now.

BarisYazici committed:
Resolves #67 update Readme

BarisYazici committed:
Resolves #67 moves colcon build to dockerfile

BarisYazici commented on ingenuity_mars_helicopter/entrypoint.sh:
```diff
@@ -0,0 +1,8 @@
+source /home/spaceros-user/spaceros/install/setup.sh
+source /opt/ros/humble/setup.sh
+cd /home/spaceros-user/ingenuity_helicopter
+colcon build --packages-select helicopter_flight_control helicopter_flight_simulation ingenuity_description ingenuity_bringup
```
I updated the Dockerfile with the colcon build command. Also adjusted the entrypoint.sh such that it directly start the launch file.

ana-GT commented:
Just as a heads-up: I gave a shot to use this PR in Jazzy. Needed a few tweaks (mainly changing ign->gz) but it seems to work great in Jazzy.  Attached a video of testing it to move Ingenuity up and down and starting a dust storm. 

https://github.com/user-attachments/assets/e9c69330-83a0-4440-a9db-7473ad2fc480.

Our tweaks are in our fork: https://github.com/traclabs/demos/tree/pr_39_ingenuity . In case this PR could use our tweaks, I am happy to send a PR to the original PR.

BarisYazici commented:
> Just as a heads-up: I gave a shot to use this PR in Jazzy. Needed a few tweaks (mainly changing ign->gz) but it seems to work great in Jazzy. Attached a video of testing it to move Ingenuity up and down and starting a dust storm.
> 
>  2025_05_10_test_ingenuity_jazzy_hb.mp4 
> 
> Our tweaks are in our fork: https://github.com/traclabs/demos/tree/pr_39_ingenuity . In case this PR could use our tweaks, I am happy to send a PR to the original PR.

That’s great @ana-GT. I am happy to hear this project is helpful for you. I guess this PR is currently stale. But it would be a good reference for the people looking for the Jazzy version.

ana-GT commented:
> That’s great @ana-GT. I am happy to hear this project is helpful for you. I guess this PR is currently stale. But it would be a good reference for the people looking for the Jazzy version.

You are welcome, thanks for adding this robot to the demos repository. I updated our fork of your PR ([this](https://github.com/traclabs/demos/tree/pr_39_ingenuity)) and it works both in Docker and building it from source. I can move the drone up and down and in the XY plane at some extent (I don't have a controller yet for moving forward/backward/left/right with some accuracy).
